src/peft/tuners/lora/layer.py

```diff
@@ -153,6 +166,65 @@ def loftq_init(self, adapter_name):
             self.lora_embedding_B[adapter_name].weight.data = lora_B
         self.get_base_layer().weight.data = qweight
 
+    def _get_weight_norm(self, weight, lora_weight, scaling) -> torch.Tensor:
+        # calculate L2 norm of weight matrix, column-wise
+        weight = weight + scaling * lora_weight
+        weight_norm = torch.linalg.norm(weight, dim=1)
+        return weight_norm
+
+    def dora_init(self, adapter_name: str) -> None:
+        lora_A = self.lora_A[adapter_name]
+        lora_B = self.lora_B[adapter_name]
+        scaling = self.scaling[adapter_name]
+        with gather_params_ctx(self.get_base_layer()):
```

Comment:
Do you think there is a way we could always avoid calling that context manager? I wonder if this might create some weird interactions if one does not use DeepSpeed Zero-3 with Trainer as `is_deepspeed_zero_3_available()` seems quite specific to `Trainer`: https://github.com/huggingface/transformers/blob/a3f9221a449e9b949e71d9b047c66186f023481f/src/transformers/integrations/deepspeed.py#L286

Reply:
I'm not sure I understand the question. We have to call the context manager or else this will break with DeepSpeed. If a user does not use DS but uses `Trainer`, should `is_deepspeed_zero3_enabled()` not return `False`? In this case the context manager just does nothing and should still work.

Reply:
ok yes makes sense!